pilot/pkg/proxy/envoy/v2/debounced_channel.go

```diff
@@ -0,0 +1,144 @@
+
+package v2
+
+import (
+	"sync"
+	"time"
+)
+
+// DebouncedEvent is used as entry type, which is enqueued in DebouncedChannel
+type DebouncedEvent = bool
+
+// DebouncedChannel is a channel which abstracts the debounce algorithm
+type DebouncedChannel struct {
+	// C channel, which can be used in selects. It delivers debounced events.
+	// The function emitted must be called. Otherwise processing is stopped.
+	C <-chan func(func(DebouncedEvent))
+	c chan func(func(DebouncedEvent))
+
+	pushCounter      uint32
+	debounceCounter  uint32
+	debounceAfterEnd time.Time
+	debounceMaxEnd   time.Time
+	lastEnqueueTime  time.Time
+	lastEmitTime     time.Time
+	debounceAfter    time.Duration
+	debounceMax      time.Duration
+	mutex            sync.Mutex
+
+	merge              func(accumulator DebouncedEvent, value DebouncedEvent) DebouncedEvent
+	accumulator        DebouncedEvent
+	initialAccumulator DebouncedEvent
+
+	running bool
+}
+
+// NewDebouncedChannel create a new instance of DebouncedChannel
+// debounceAfter: A new outgoing event is emitted if no additional event is enqueued during this period (resets on every enqueued event).
+// debounceMax: A new outgoing event is emitted after this period (even if additional events were enqueued in that period).
+// initialAccumulator: Initial value of the accumulator
+// merge: function, which is called to merge two events if they are debounced.
+func NewDebouncedChannel(debounceAfter time.Duration, debounceMax time.Duration,
+	initialAccumulator DebouncedEvent, merge func(accumulator DebouncedEvent, value DebouncedEvent) DebouncedEvent) *DebouncedChannel {
+
+	c := make(chan func(func(DebouncedEvent)), 1)
+
+	return &DebouncedChannel{
+		C:                  c,
+		c:                  c,
+		initialAccumulator: initialAccumulator,
+		accumulator:        initialAccumulator,
+		merge:              merge,
+		debounceAfter:      debounceAfter,
+		debounceMax:        debounceMax,
+		lastEmitTime:       time.Unix(0, 0),
+	}
+
+}
+
+// Enqueue pushes a new event to the channel
+func (dc *DebouncedChannel) Enqueue(element DebouncedEvent) {
+	dc.mutex.Lock()
+	defer dc.mutex.Unlock()
+	dc.accumulator = dc.merge(dc.accumulator, element)
+	if dc.debounceCounter == 0 {
+		dc.debounceMaxEnd = time.Now().Add(dc.debounceMax)
+		go func() {
+			dc.mutex.Lock()
+			defer dc.mutex.Unlock()
+
+			for {
+				debounceCounter := dc.debounceCounter
+
+				sleepDuration := min(dc.debounceAfter, time.Until(dc.debounceMaxEnd))
+				dc.mutex.Unlock()
+				time.Sleep(sleepDuration)
+				dc.mutex.Lock()
+
+				if debounceCounter == dc.debounceCounter || time.Now().After(dc.debounceMaxEnd) {
+					break
+				}
+			}
+			dc.runIfNeeded()
+		}()
+	}
+
+	dc.debounceAfterEnd = time.Now().Add(dc.debounceAfter)
+	dc.debounceCounter++
+
+	dc.lastEnqueueTime = time.Now()
+
+	dc.runIfNeeded()
+}
+
+func min(a, b time.Duration) time.Duration {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func (dc *DebouncedChannel) runIfNeeded() {
```

Comment:
a more meaningful name?

Reply:
What about `runIfNeededAndPossible`?

`runIfADebounceTimersHasRunOutAndAJobIsPendingAndNoOtherJobIsCurrentlyRunning` would be to long i guess :smile:

See https://github.com/istio/istio/pull/15810/files/513e1ea0447323a7ae0190fc279acb3068c0e5b9#r308660111 for more details